source-academy/frontend
# Remove editor ref forwarding in favour of callbacks

### Description

The mobile workspace comes with a mobile keyboard for inputting characters common to Source that might be hard to type on typical phone keyboards. To allow the mobile keyboard to manipulate the editors (which are child components of the `MobileWorkspace` component), the current codebase makes use of React refs to track which editor has focus:
https://github.com/source-academy/frontend/blob/78ff16511385a31e8e66fbfd72a3e201cbbbe8bc/src/commons/mobileWorkspace/MobileWorkspace.tsx#L95-L112

This however results in the need to forward refs through the component hierarchy from the child editor components to the `MobileWorkspace` component which is rather messy. In two separate locations, the merging of refs is used, which further complicates the codebase.

Instead of having the parent `MobileWorkspace` component be responsible for setting the event listeners on the child editor components (as seen in the code snippet above), we can make use of callbacks which are passed to the child editor components. This shifts the responsibility of maintaining event listeners to the child editor components, resulting in more cohesive code. On top of that, this also removes the need to merge refs, leading to more readable code.

Part of the refactoring for #2176 in preparation for adding multiple editor tabs.

### Type of change


### How to test

Go into the mobile view and test that the mobile keyboard is able to input characters into the playground editor, the REPL, and the Sourcecast editor (the playground editor & Sourcecast editor are different components). Note that there is a separate issue where the mobile keyboard steals focus from the underlying editor (see #2276).
